cdap-unit-test/src/main/java/io/cdap/cdap/test/LogSystemExitRule.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.test;
+
+import org.junit.rules.ExternalResource;
+
+import java.security.Permission;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * <p>
+ *   For any {@link System#exit(int)} call logs it to the stderr along with stack trace.
+ *   This rule is useful to catch cases with
+ *   "The forked VM terminated without properly saying goodbye. VM crash or System.exit called?"
```

Comment:
I don't think there is place we actually calls `System.exit`, will this SM be useful if the VM quite because of OOM (we have `-XX:OnOutOfMemoryError="kill -9 %p"` in the surefire config)

Reply:
I was considering this, but I am not sure. I was checking logs and usually when there is an OOM there is at least a log line about it. This change should give more information on if that's OOM (due to Java OOM of even linux process killer) or some `System.exit`